src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -234,12 +225,118 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
+            // we still believe this is acceptable given how critical perf is in typing scenario.
+            // Additionally, expanded items are usually considered complementary. The need for them only rise occasionally (it's rare when users need to add imports,)
+            // and when they are needed, our hypothesis is because of their more intrusive nature (adding an import to the document) users would more likely to
+            // contemplate such action thus typing slower before commit and/or spending more time examining the list, which give us some opportunities
+            // to still provide those items later before they are truly required.            
+
+            if (!options.ShouldShowItemsFromUnimportNamspaces())
+            {
+                // No need to trigger expanded providers at all if the feature is disabled, just trigger core providers and return;
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+                AddPropertiesToSession(session, list, triggerLocation);
+                return context;
+            }
+
+            if (!session.TextView.Options.GetOptionValue(DefaultOptions.ResponsiveCompletionOptionId))
+            {
+                // We tie the behavior of delaying expand items to editor's "responsive completion" option.
+                // i.e. "responsive completion" disabled == always wait for all items to be calculated.
+                options = options with
+                {
+                    BlockOnExpandedCompletion = true,
+                    ExpandedCompletionBehavior = ExpandedCompletionMode.AllItems,
+                };
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation, options, cancellationToken).ConfigureAwait(false);
+                AddPropertiesToSession(session, list, triggerLocation);
+                return context;
+            }
+
+            var expandedItemsTask = Task.Run(async () => await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                options with { ExpandedCompletionBehavior = ExpandedCompletionMode.ExpandedItemsOnly }, cancellationToken).ConfigureAwait(false),
+                cancellationToken);
+
+            // Record how long it takes for the background task to complete.
+            var stopwatch = Stopwatch.StartNew();
+            expandedItemsTask = expandedItemsTask.ContinueWith(task =>
+            {
+                AsyncCompletionLogger.LogAdditionalTicksToCompleteDelayedImportCompletionDataPoint((int)stopwatch.ElapsedMilliseconds);
+                return task.Result;
+            }, cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+
```

Comment:
nit. t he following code is so complex that i woudl take the main three scenarios and write 3 different helper methods that this method calls.

Reply:
this still applies.  i would at least break this code out.